### DISCUSSION TITLE
Auto-import sbt projects

### DISCUSSION THREAD

**Discussion by fdietze** [scalameta/metals#3343] [Q&A]
Hi,

is there a way to always auto-import the project, instead of getting prompted after changing the `build.sbt`?

When editing quickly, sometimes the prompt gets in the way.

Might be related to #328 

Thank you!

**Comment by ckipp01** [comment]
Hey @fdietze! Good question, we actually by default don't do this. The full import is actually pretty expensive, so try to not hammer the user with requests if they are making a bunch of changes one right after the other. Typically in a senario like this I just say "Not now" which will silent it for a while, and then manually trigger the build import when I'm done.

**Reply by fdietze** [reply]
Thanks for your answer. For small changes in the `build.sbt`, isn't most of the build cached anyways? I'd be fine with having the import run everytime I change the `build.sbt`. That's better than getting interrupted by the prompt.

**Reply by ckipp01** [reply]
> isn't most of the build cached anyways

Every time you do an import (when using Bloop), you need to run a full `bloopInstall` process. This is expensive.

> That's better than getting interrupted by the prompt.

I'd recommend just answering "Not Now", which will stop the prompts until you want to explicitly do the build import.

**Reply by fdietze** [reply]
I already do select "not now", but it's still interrupting me. Very often while typing and then I type text into the prompt.

**Reply by ckipp01** [reply]
Have you tried the "Don't show again" option then? And then just manually trigger it every time? Again, by design there isn't a way to do this were you auto-import at all time. It'd be a pretty awful user experience.

**Reply by fdietze** [reply]
I tried all the options and nothing was working well for me so far... I don't want to see the prompt at all, but still use metals. If it's doing something expensive in the background when I'm working on `build.sbt`, that's fine for me.

**Reply by cornerman** [reply]
I would also like to have an auto-import functionality. With the current setup, the prompt interrupts my work in the editor and then I anyhow always press yes.

**Reply by ckipp01** [reply]
I transfered this over to the main metals repo since this isn't really specific to `coc-metals` and all of the extensions will behave this way since that's how Metals works.

> I tried all the options and nothing was working well for me so far... I don't want to see the prompt at all,

So choosing the "Don't show again" option the first time it appears should work like this. Once you choose this it will no longer show you the prompt.

**Reply by dos65** [reply]
We can't make auto-import as it is a slow process and there is no way to make it work fast enough to provide auto-import per every save in build file.

I'm wondering if can disable this notification?
I never use it and always perform import by a command.

**Reply by tgodzik** [reply]
> I'm wondering if can disable this notification?
> I never use it and always perform import by a command.

This might be useful for beginners, so I think `Don't show again` is a good enough option for more experienced users.

As for auto-import, we could actually have it for sbt BSP server. I think that should work alright if we check before for any diagnostics.
